Add tab for Zarr metadata

charlesbluca opened the pull request:
This PR adds a tab to view the metadata of a consolidated Zarr data store, conditional on the existence of an asset with the `"zarr-consolidated-metadata"` role, as discussed in #44:

![screenshot of Zarr metadata tab in action](https://user-images.githubusercontent.com/20627856/97036770-509fcf80-1536-11eb-9f8d-9b6821b34e44.png)

This introduces the following changes:

- Adds [axios](https://www.npmjs.com/package/axios) to dependencies in [package.json](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/package.json#L19) to perform the GET request of the metadata file
- Adds `zarrMetadataUrl` as a computed property in [common.js](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/common.js#L279) returning the `href` of the appropriate asset, if it exists
- Adds `ZarrMetadataTab` component to [Catalog.vue](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/Catalog.vue#L128) to be visible and active on the existence of a `zarrMetadataUrl`
- Creates [ZarrMetadataTab.vue](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/ZarrMetadataTab.vue) to handle the downloading, parsing, and rendering of the Zarr metadata

If present, the tab represents each array of the Zarr data store as an item in a Bootstrap list group, with a table listing its dimensions and their corresponding shape/chunk size, and a collapsible card containing additional metadata:

![screenshot of array attributes card](https://user-images.githubusercontent.com/20627856/97035810-e20e4200-1534-11eb-8109-4055c952a0e4.png)


Below the list of arrays is a similar card containing all group-level attributes:

![screenshot of group attributes card](https://user-images.githubusercontent.com/20627856/97035739-c5720a00-1534-11eb-815f-92bcf20d5b88.png)

An example of a Collection with accessible Zarr metadata (which I used for testing) can be found at:

- https://raw.githubusercontent.com/pangeo-data/pangeo-datastore-stac/master/test_zarr/collection.json

An example of a Collection with _inaccessible_ Zarr metadata (due to CORS headers not being set in the containing bucket) can be found at:

- https://raw.githubusercontent.com/pangeo-data/pangeo-datastore-stac/master/master/ocean/sea_surface_height/collection.json

In the case of an error, a message appears where the arrays and attributes would have normally been. I left this intentionally vague for now in case a better solution is recommended (disabling the tab, placing the error message elsewhere on the page, etc.):

![screenshot of a sample error message](https://user-images.githubusercontent.com/20627856/97036615-16cec900-1536-11eb-905e-f73b74b0794d.png)

Other than that, some other factors to consider:

- Should any more properties be present for each Zarr array?
- Are there any properties which could be laid out in a different fashion?
- Can the styling be altered to better mesh with other elements of the STAC browser?

m-mohr commented:
This is looking great and it's nice to see additional formats being supported in STAC Browser.

I have two remarks:
1. With the introduction of axios there are now two ways to perform GET requests (see fetchUri in utils). There should be only one. So either use fetchUri or use axios everywhere.
2. I think it would help loading times to have the ZarrMetadataTab included as a async Vue component, see https://vuejs.org/v2/guide/components-dynamic-async.html#Async-Components the `'my-component': () => import('./my-async-component')` example. Most times this component is not needed so it should only be loaded once required by the data.

m-mohr commented on src/components/common.js:
```diff
@@ -275,6 +275,16 @@ export default {
+        x.roles.includes("zarr-consolidated-metadata")
+      );
+
+      if (zarrMetadata != null) {
```
Arra.find returns undefined if no result is found, not null.

```suggestion
      if (typeof zarrMetadata !== 'undefined') {
```

m-mohr commented on src/components/Catalog.vue:
```diff
@@ -207,7 +213,7 @@ export default {
     }
   },
-  components: { AssetTab, MetadataSidebar },
+  components: { AssetTab, ZarrMetadataTab, MetadataSidebar },
```
Load component on demand only:

```suggestion
  components: {
    AssetTab,
    ZarrMetadataTab: () => import('./ZarrMetadataTab.vue'),
    MetadataSidebar
  },
```

charlesbluca replied:
This change works as intended, but seems to create a parsing error with ESLint that prevents it from linting the rest of the file - is this something we should work out before pushing this change?

A seemingly similar issue being dealt with here:

https://stackoverflow.com/questions/47815775/dynamic-imports-for-code-splitting-cause-eslint-parsing-error-import

m-mohr commented on src/components/Catalog.vue:
```diff
@@ -172,6 +177,7 @@ import { mapActions, mapGetters } from "vuex";
 
 import common from "./common";
 import AssetTab from './AssetTab.vue'
+import ZarrMetadataTab from "./ZarrMetadataTab.vue";
```
```suggestion
```

Remove this line, see changes below.

m-mohr commented on src/components/ZarrMetadataTab.vue:
```diff
@@ -0,0 +1,150 @@
+<script>
+const axios = require("axios").default;
+
+function parseMeta(meta) {
```
Why is this outside of the components and not inside the components methods?

charlesbluca replied:
This was an oversight on my part - I'll move this to component methods.

m-mohr reviewed (changes requested):
This is looking great and it's nice to see additional formats being supported in STAC Browser.

I have two remarks:
1. With the introduction of axios there are now two ways to perform GET requests (see fetchUri in utils). There should be only one. So either use fetchUri or use axios everywhere.
2. I think it would help loading times to have the ZarrMetadataTab included as a async Vue component, see https://vuejs.org/v2/guide/components-dynamic-async.html#Async-Components the `'my-component': () => import('./my-async-component')` example. Most times this component is not needed so it should only be loaded once required by the data.

charlesbluca commented:
Thanks for the review, @m-mohr! I'll switch over to `fetchUri` instead of axios for the GET request. Additionally, I now know that there are several aspects of this metadata view that should be expanded/altered to offer a "complete" view of generic Zarr data stores:

- Adding support for nested Zarr groups - essentially, this would entail making the [list group](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/ZarrMetadataTab.vue#L9) and its contents a template which could be called recursively on groups within the root group structure
- Making the collapsible attributes cards optional, as group/array level attributes are not required in a Zarr store
- Render array dimension names conditional to the existence of the [_ARRAY_DIMENSIONS] attribute, which is not required in a Zarr store

I'll keep this PR updated with progress on these changes.

m-mohr commented on src/components/common.js:
```diff
@@ -281,7 +281,7 @@ export default {
       );
 
-      if (typeof zarrMetadata !== 'undefined') {
+      if (typeof zarrMetadata !== undefined) {
```
The quotes were correct 🤔

charlesbluca replied:
My bad, was debugging this one in console and totally missed the `typeof` - I'll change this back.

charlesbluca commented:
I've now added the following changes:

- Removed axios from the dependencies, now relying on preexisting dependency [node-fetch](https://www.npmjs.com/package/node-fetch) to perform the GET request
- Set [ZarrMetadataTab to be an async component](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/Catalog.vue#L217) with help from @m-mohr (also reset its active status to false)
- Altered the check for a Zarr metadata URL in [common.js](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/common.js#L284) to better reflect the output of `Array.prototype.find()`
- Moved [`parseMeta`](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/ZarrMetadataTab.vue#L39) to the component methods of ZarrMetadataTab and expanded the method to handle nested Zarr groups
- Created a separate [ZarrGroup](https://github.com/charlesbluca/stac-browser/blob/add-zarr-tab/src/components/ZarrGroup.vue) component to be used recursively in the case of nested Zarr groups, with a couple of changes from the original component's implementation:
  - Uses preexisting dependency [lodash.isempty](https://lodash.com/docs/4.17.15#isEmpty) to perform checks for if array/group level attributes should be rendered
  - Shows dimension names in table _only_ if array has attribute `_ARRAY_DIMENSIONS`; otherwise uses indexed axes
  - Shows dtype as unaltered string from metadata file - though I would like more descriptive names such as `int16` or `float32`, I don't know all the shorthand names for the [NumPy data types](https://numpy.org/doc/stable/user/basics.types.html) these are derived from

In sum this allows potentially dimensionless, attribute-less, nested Zarr groups/arrays to be rendered without a problem, although this can result in a much busier view of the metadata:

![Screenshot of doubly nested Zarr array](https://user-images.githubusercontent.com/20627856/97498212-28530f00-1942-11eb-84a5-f61542571d32.png)

If this isn't a problem, then I don't think there is much more functionality that needs to be added here to render any generic Zarr group. In the long term, I would like to use styling to make a cleaner view, maybe reminiscent of xarray's HTML dataset preview:

![Screenshot of xarray dataset preview](https://user-images.githubusercontent.com/20627856/97501919-84209680-1948-11eb-8b3f-bfc991c01f28.png)

m-mohr reviewed (approved):
Thanks for tackling my requests. I haven't looked into details again and have not checked changes in STAC Index, but I guess it should work. So approving for now :-)

lossyrob commented:
Planning on taking a look at this soon - excited to see this contribution!

rabernat commented on src/components/common.js:
```diff
@@ -275,6 +275,16 @@ export default {
+    },
+    zarrMetadataUrl() {
+      const zarrMetadata = this.assets.find(x =>
+        x.roles.includes("zarr-consolidated-metadata")
```
```suggestion
        x.roles.includes("zarr-v2-consolidated-metadata")
```

I think we should use a version-specific attribute here. Zarr v3 will look different.

charlesbluca replied:
I agree with having version-specific parsing/rendering of the metadata - if we plan on supporting Zarr v3 down the line, we should also figure out how we would like to pass information about the Zarr version down to the `parseMeta` function:

- This computed property could return a version number along with the URL based on the role of the Zarr metadata asset
- The `ZarrMetadataTab` could check the version by accessing the group's `zarr_format` after downloading the file

In either case, this version could be used to call a specified `parseMetaVX`, or throw an error if the version isn't supported.

lossyrob reviewed (approved):
This looks great!

As the PR description suggests, I think there's probably some styling updates that would integrate it better with the UI - however I don't believe this is blocking and can be handled in follow up changes.

Also, the issue brought up with the versioned role discussion can be addressed in a follow up - I'd recommend writing an issue if there's more discussion to have about what the specific role text should be.

Merging!

rabernat commented:
Thanks to @charlesbluca and the stac-browser devs for supporting this PR. It will really help us migrate our Pangeo catalogs to STAC!

> I think there's probably some styling updates that would integrate it better with the UI

Any ideas who might be able to help with this? We don't have a lot of designers in our network.

m-mohr commented:
If you could fix the CORS issue somehow, it would also show up on STAC Index  now: https://stacindex.org/catalogs/pangeo#/2VtvSQ1Nw1ZcCeDgFRPUHSLLsouLBcQMmyDep3mN1ZK7LLg6E85XEm5j9A8MraeQpNLTzaqAA6yBfm6tYeJz7e1YAoQgHxJKVSLKMcUrV6d/2o6asgyf1uitBnbjefHbHK3nWhmPxzSXDKQktRiGed4uUUpka6mQDxDhNFnU1VDEbHjgySTxHwXrTeGUvE9PE98YhjbV9VwEqytToTqTqLWmGQ5MmDXSRTfs2UuvNfu4e75PfXT9f?t=summaries

lossyrob commented:
> Any ideas who might be able to help with this? We don't have a lot of designers in our network.

@rabernat I don't, at least for volunteers in the STAC network. Perhaps we could try to attract some design resources to the next STAC sprint to work on making the site look better overall, and visually tying everything together in the UI.

m-mohr commented:
@charlesbluca FYI: I quickly had to address a bug introduced here that made Items without roles in assets not render any longer: https://github.com/radiantearth/stac-browser/commit/69e239832207abebc352ba2972de153b00dc7e0b
